# CODE FOR GIVEN ALGORITHMS IN RESEARCH PAPER

MAKING THE DATASET FROM GIVEN FILES

Dataset: 

In the research paper link for the dataset is given, which contain three folder as P,Q,N. these folder contains many files of theory releated to families of protin.
In folder P there are files which belongs to positive class means releated to transport protin, in folder Q file also belongs to
positive class and in N all files belongs to negative class means they are not related to transport protin.
for P i made S=1 which means they are labelled and for U(Q union N) S=-1 means they are unlabelled.
Our aim is that made a classifer using labelled and unlabelled data which classify positive and negative class for a datapoint.



In [321]:
import pandas as pd
import os

folder_path = 'C:/Users/nitin_2n4diox/Downloads/P'
dataframes_list = []

for file_name in os.listdir(folder_path):
    file_path = os.path.join(folder_path, file_name)
    with open(file_path, 'r') as file:
        text = file.read()
        dataframes_list.append(pd.DataFrame([str(text)], columns=['Text'], dtype=str))

df1 = pd.concat(dataframes_list, ignore_index=True, axis=0)



In [322]:
df1

Text
0     AC   P0A770; P77145;\nManganese transport prot...
1     AC   P0A769; P77145;\nManganese transport prot...
2     AC   P37329;\nMolybdate-binding periplasmic pr...
3     AC   P0AF02; P09834; P77537;\nMolybdenum trans...
4     AC   P0AF01; P09834; P77537;\nMolybdenum trans...
...                                                 ...
2394  AC   P01501; P01503;\nMelittin precursor (Alle...
2395  AC   P40260;\nAmmonium transporter MEP1.\n"Clo...
2396  AC   P41948;\nAmmonium transporter MEP2.\n"A f...
2397  AC   P30344;\nMercury resistance probable Hg t...
2398  AC   Q10419;\nMesentericin Y105 secretion prot...

[2399 rows x 1 columns]

In [323]:
df1['actual_label(Y)']='1'
df1['S']='1'

In [324]:
df1

Text actual_label(Y)  S
0     AC   P0A770; P77145;\nManganese transport prot...               1  1
1     AC   P0A769; P77145;\nManganese transport prot...               1  1
2     AC   P37329;\nMolybdate-binding periplasmic pr...               1  1
3     AC   P0AF02; P09834; P77537;\nMolybdenum trans...               1  1
4     AC   P0AF01; P09834; P77537;\nMolybdenum trans...               1  1
...                                                 ...             ... ..
2394  AC   P01501; P01503;\nMelittin precursor (Alle...               1  1
2395  AC   P40260;\nAmmonium transporter MEP1.\n"Clo...               1  1
2396  AC   P41948;\nAmmonium transporter MEP2.\n"A f...               1  1
2397  AC   P30344;\nMercury resistance probable Hg t...               1  1
2398  AC   Q10419;\nMesentericin Y105 secretion prot...               1  1

[2399 rows x 3 columns]

In [325]:

folder_path = 'C:/Users/nitin_2n4diox/Downloads/Q'
dataframes_list = []

for file_name in os.listdir(folder_path):
    file_path = os.path.join(folder_path, file_name)
    with open(file_path, 'r') as file:
        text = file.read()
        dataframes_list.append(pd.DataFrame([str(text)], columns=['Text'], dtype=str))

df2 = pd.concat(dataframes_list, ignore_index=True, axis=0)


In [326]:
df2['actual_label(Y)']='1'
df2['S']='-1'

In [327]:
df2

Text actual_label(Y)   S
0    AC   P46492;\nProbable D-methionine transport ...               1  -1
1    AC   Q97TN5;\nProbable manganese transport pro...               1  -1
2    AC   P57788; Q9DEY5; Q9DG24;\nMonocarboxylate ...               1  -1
3    AC   O61703;\nPhosphate carrier protein, mitoc...               1  -1
4    AC   Q9C8H0;\nMultidrug resistance-associated ...               1  -1
..                                                 ...             ...  ..
343  AC   P13345;\nLysis protein for colicin E6 pre...               1  -1
344  AC   P02911;\nLysine-arginine-ornithine-bindin...               1  -1
345  AC   Q89A97;\nMultidrug resistance-like ATP-bi...               1  -1
346  AC   O80725;\nMultidrug resistance protein 4 (...               1  -1
347  AC   Q8FAX1;\nMultidrug resistance protein mdt...               1  -1

[348 rows x 3 columns]

In [328]:
folder_path = 'C:/Users/nitin_2n4diox/Downloads/N'
dataframes_list = []

for file_name in os.listdir(folder_path):
    file_path = os.path.join(folder_path, file_name)
    with open(file_path, 'r') as file:
        text = file.read()
        dataframes_list.append(pd.DataFrame([str(text)], columns=['Text'], dtype=str))

df3 = pd.concat(dataframes_list, ignore_index=True, axis=0)


In [329]:
df3['actual_label(Y)']='-1'
df3['S']='-1'

In [330]:
df3

Text actual_label(Y)   S
0     AC   Q20YV7;\nS-adenosylmethionine synthetase ...              -1  -1
1     AC   Q5FSK6;\nS-adenosylmethionine synthetase ...              -1  -1
2     AC   Q2RK28;\nS-adenosylmethionine synthetase ...              -1  -1
3     AC   P78003;\nS-adenosylmethionine synthetase ...              -1  -1
4     AC   Q9RL99; Q68VX2;\nS-adenosylmethionine syn...              -1  -1
...                                                 ...             ...  ..
4553  AC   Q8CU52;\nAlkylmercury lyase (EC 4.99.1.2)...              -1  -1
4554  AC   Q5R6F1;\nMesoderm development candidate 2...              -1  -1
4555  AC   P44502;\nCystathionine gamma-synthase (EC...              -1  -1
4556  AC   P66875; O53427;\nCystathionine gamma-synt...              -1  -1
4557  AC   Q6A816;\nArgininosuccinate lyase (EC 4.3....              -1  -1

[4558 rows x 3 columns]

In [331]:
#CONCAT ALL DATA 
df= pd.concat([df1, df2, df3], axis=0,sort=False)

In [332]:
df

Text actual_label(Y)   S
0     AC   P0A770; P77145;\nManganese transport prot...               1   1
1     AC   P0A769; P77145;\nManganese transport prot...               1   1
2     AC   P37329;\nMolybdate-binding periplasmic pr...               1   1
3     AC   P0AF02; P09834; P77537;\nMolybdenum trans...               1   1
4     AC   P0AF01; P09834; P77537;\nMolybdenum trans...               1   1
...                                                 ...             ...  ..
4553  AC   Q8CU52;\nAlkylmercury lyase (EC 4.99.1.2)...              -1  -1
4554  AC   Q5R6F1;\nMesoderm development candidate 2...              -1  -1
4555  AC   P44502;\nCystathionine gamma-synthase (EC...              -1  -1
4556  AC   P66875; O53427;\nCystathionine gamma-synt...              -1  -1
4557  AC   Q6A816;\nArgininosuccinate lyase (EC 4.3....              -1  -1

[7305 rows x 3 columns]

In [333]:
#to make text from pandas.core.Series.series
df.to_csv('file3.csv', index=False)

In [334]:
DF = pd.read_csv('file3.csv')
DF

Text  actual_label(Y)  S
0     AC   P0A770; P77145;\nManganese transport prot...                1  1
1     AC   P0A769; P77145;\nManganese transport prot...                1  1
2     AC   P37329;\nMolybdate-binding periplasmic pr...                1  1
3     AC   P0AF02; P09834; P77537;\nMolybdenum trans...                1  1
4     AC   P0AF01; P09834; P77537;\nMolybdenum trans...                1  1
...                                                 ...              ... ..
7300  AC   Q8CU52;\nAlkylmercury lyase (EC 4.99.1.2)...               -1 -1
7301  AC   Q5R6F1;\nMesoderm development candidate 2...               -1 -1
7302  AC   P44502;\nCystathionine gamma-synthase (EC...               -1 -1
7303  AC   P66875; O53427;\nCystathionine gamma-synt...               -1 -1
7304  AC   Q6A816;\nArgininosuccinate lyase (EC 4.3....               -1 -1

[7305 rows x 3 columns]

In [335]:
type(DF['Text'][0])
type(DF['S'][0])

numpy.int64

PREPROCESSING THE TEXT DATASEET AND CONVERTING IN TO VECTOR USING TF-IDF.

In [336]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

In [337]:
messages=DF.copy()
messages.head()


Text  actual_label(Y)  S
0  AC   P0A770; P77145;\nManganese transport prot...                1  1
1  AC   P0A769; P77145;\nManganese transport prot...                1  1
2  AC   P37329;\nMolybdate-binding periplasmic pr...                1  1
3  AC   P0AF02; P09834; P77537;\nMolybdenum trans...                1  1
4  AC   P0AF01; P09834; P77537;\nMolybdenum trans...                1  1

In [338]:
type(messages['Text'][0])

str

In [339]:
len(messages)

7305

In [340]:
##text preprocessing
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
import re
ps = PorterStemmer()
corpus = []
for i in range(0, len(messages)):
    review = re.sub('[^a-zA-Z]', ' ', messages['Text'][i])
    review = review.lower()
    review = review.split()
    
    review = [ps.stem(word) for word in review if not word in stopwords.words('english')]
    review = ' '.join(review)
    corpus.append(review)

In [341]:
corpus[40]

'ac p c p small conduct mechanosensit channel genom sequenc shigella flexneri insight pathogen comparison genom escherichia coli k complet genom sequenc compar genom shigella flexneri serotyp strain function mechanosensit channel particip regul osmot pressur chang within cell form ion channel nanosiemen conduct slight prefer anion channel sensit voltag similar subunit homoheptam similar subcellular locat cell inner membran multi pass membran protein similar similar belong msc tc famili complet proteom inner membran ion transport ionic channel membran transmembran transport ththree'

In [342]:
## TFidf Vectorizer
from sklearn.feature_extraction.text import TfidfVectorizer
tfidf_v=TfidfVectorizer(max_features=5000,ngram_range=(1,3))
X=tfidf_v.fit_transform(corpus).toarray()

In [343]:
X.shape

(7305, 5000)

# 1. Ideal : Training on positive vs negative datapoints

In [344]:
y=messages['actual_label(Y)']

In [345]:
## Divide the dataset into Train and Test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X,y , test_size=0.30, random_state=0)

In [346]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

# create an instance of the SVM classifier with a linear kernel and a soft margin
clf = SVC(kernel='linear', C=1.0, probability=True)

# train the classifier on the training data
clf.fit(X_train, y_train)

# make predictions on the test data
y_pred = clf.predict(X_test)

# calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# print the confusion matrix
print(cm)

[[1327   35]
 [  45  785]]


In [347]:
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, roc_curve



# Calculate accuracy
accuracy = accuracy_score(y_test, y_pred)

# Calculate F1 score
f1 = f1_score(y_test, y_pred)

# Calculate AUC-ROC score
auc_roc = roc_auc_score(y_test, y_pred)

# Get ROC curve values
#fpr, tpr, thresholds = roc_curve(actual_label, f_pred)

print("Accuracy:", accuracy)
print("F1 score:", f1)
print("AUC-ROC score:", auc_roc)

Accuracy: 0.9635036496350365
F1 score: 0.9515151515151515
AUC-ROC score: 0.9600428144295243


Conclusion: 
this is simple classification problem using linear soft-svm , so here i got good result as 
Accuracy: 0.9635036496350365
F1 score: 0.9515151515151515
AUC-ROC score: 0.9600428144295243 
it is almost equal to given research paper result.

# 2. Using probablistic approach

In [348]:
S=messages['S']

In [349]:
## Divide the dataset into Train and Test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, S, test_size=0.30, random_state=0)

In [350]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

# create an instance of the SVM classifier with a linear kernel and a soft margin
clf = SVC(kernel='linear', C=1.0, probability=True)

# train the classifier on the training data
clf.fit(X_train, y_train)

# make predictions on the test data
y_pred = clf.predict(X_test)

# calculate the confusion matrix
cm = confusion_matrix(y_test, y_pred)

# print the confusion matrix
print(cm)

# predict the probability of each test data point belonging to class 1
#proba = clf.predict_proba(x_test)[:, 1]
def Nontraditional_classifier(X_test):
    return clf.predict_proba(X_test)[:, 1]
    
# print the probabilities
#print(proba)


[[1377   97]
 [  65  653]]


In [351]:
##for finding the expected value of C= average of Nontraditional classifier on labelled datapoints.
c=0
for i in range (2399):
    p=X[i]
    p=p.reshape(1,-1)
    c+= Nontraditional_classifier(p)
C=c/2399
print("Expected value of nontraditional classifier on positive datapoints C =", C)
    

Expected value of nontraditional classifier on positive datapoints C = [0.86073617]


Here i got the value of C is 0.86073617 which is almost equal to the value obtained in research paper (0.86073617)

In [352]:

p=X[1567]
p=p.reshape(1,-1)
l= Nontraditional_classifier(p)
print(l)

[0.97682125]


In [353]:
def Traditional_classifier(X_test):
    g=Nontraditional_classifier(X_test)
    f=g/C
    return f

    

In [358]:
##to find the predicted values such that the datapoints are belongs to postive or negative class (y=1 or 0).
import numpy
lst=[]

for z in range(7305):
    p=X[z]
    p=p.reshape(1,-1)
    l=Traditional_classifier(p)
    if l>= 0.5:
        m=1
    else:
        m=-1
    lst.append(m)

f_pred = numpy.array(lst)

In [359]:
new_array=DF['actual_label(Y)']

In [360]:

actual_label = new_array.to_numpy()
type(actual_label)

numpy.ndarray

In [361]:
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, roc_curve



# Calculate accuracy
accuracy = accuracy_score(actual_label, f_pred)

# Calculate F1 score
f1 = f1_score(actual_label, f_pred)

# Calculate AUC-ROC score
auc_roc = roc_auc_score(actual_label, f_pred)

# Get ROC curve values
#fpr, tpr, thresholds = roc_curve(actual_label, f_pred)

print("Accuracy:", accuracy)
print("F1 score:", f1)
print("AUC-ROC score:", auc_roc)


Accuracy: 0.9561943874058864
F1 score: 0.939209726443769
AUC-ROC score: 0.945009019373003


Conclusion: 
in this section i made traditional and nontraditional classifier using only labelled and unlabelled datapoints. this is main algorithm of paper. here i got expected value of C = 0.86073617 which is almost equal to paper value.
Accuracy: 0.9561943874058864
F1 score: 0.939209726443769
AUC-ROC score: 0.945009019373003
finally these result match with research paper result.

# 3. Weighting approach

In [362]:
## As in research paper we know that W is defined as p(y=1|x) and for this formula is derived in paper.

def weight(X_test):
    g=Nontraditional_classifier(X_test)
    g=g[0]
    C=0.85960346
    w=((1-C)/C)*(g/(1-g))
    return w

In [365]:
lst2=[]
for q in range(7305):
    if q<=2398:
        m=1
    else:
        p=X[q]
        p=p.reshape(1,-1)
        l=weight(p)
        if l>= 0.5:
            m=1
        else:
            m=-1
    lst2.append(m)

weight_pred = numpy.array(lst2)

In [366]:
weight_pred
type(weight_pred)

numpy.ndarray

In [367]:
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, roc_curve



# Calculate accuracy
accuracy = accuracy_score(actual_label, weight_pred)

# Calculate F1 score
f1 = f1_score(actual_label, weight_pred)

# Calculate AUC-ROC score
auc_roc = roc_auc_score(actual_label, weight_pred)

# Get ROC curve values
#fpr, tpr, thresholds = roc_curve(actual_label, f_pred)

print("Accuracy:", accuracy)
print("F1 score:", f1)
print("AUC-ROC score:", auc_roc)

Accuracy: 0.9670088980150582
F1 score: 0.9542434023163091
AUC-ROC score: 0.9566402008940944


Conclusion:
As the derivation mentioned in paper for calculate the weight p(y=1|x), i made function using Nontraditional classifier and value C.
Accuracy: 0.9670088980150582
F1 score: 0.9542434023163091
AUC-ROC score: 0.9566402008940944
These are equivalent to paper result.

# 4. Biased SVM from previous reasearch paper.

In [368]:
y=messages['actual_label(Y)']

In [369]:
S=messages['S']

In [370]:
## Divide the dataset into Train and Test
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, S, test_size=0.30, random_state=0)

In [371]:
from sklearn.svm import SVC
from sklearn.metrics import confusion_matrix

# create an instance of the SVM classifier with a linear kernel and a soft margin
clf = SVC(kernel='linear', probability=True,class_weight={1:0.1,-1:0.01})

# train the classifier on the training data
clf.fit(X_train, y_train)

# make predictions on the test data
y_pred = clf.predict(X_test)




In [372]:
# In above cell y_pred is telling the predected values of S(labelled or unlabelled) but we will consider these values
# as of y(positive or negative classified) as mentioned in paper. And compare these with actual label


from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y , test_size=0.30, random_state=0)

corresponding_y = y_test.to_numpy()



cm = confusion_matrix(corresponding_y, y_pred)

# print the confusion matrix
print(cm)

[[ 319 1043]
 [   0  830]]


In [374]:
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, roc_curve



# Calculate accuracy
accuracy = accuracy_score(corresponding_y, y_pred)

# Calculate F1 score
f1 = f1_score(corresponding_y, y_pred)

# Calculate AUC-ROC score
auc_roc = roc_auc_score(corresponding_y, y_pred)

# Get ROC curve values
#fpr, tpr, thresholds = roc_curve(actual_label, f_pred)

print("Accuracy:", accuracy)
print("F1 score:", f1)
print("AUC-ROC score:", auc_roc)

Accuracy: 0.5241788321167883
F1 score: 0.6141324454310025
AUC-ROC score: 0.6171071953010279


Conclusioin:
This approach was given in previous study. here i gave the different-2 value of c in svm classifier for labelled and unlabelled datapoints.
Results are not good because we know that in case of noisy dataset( in our case negative points are noisy with positive one) it will not work well. 
But as we know that by this algo. we got high TPR, so it is useful in such scenario where high TPR required like in medical dataset(identified cancer patients).

# Novelty (clustering)

In [375]:
from sklearn.cluster import KMeans



# initialize k-means with 2 clusters
kmeans = KMeans(n_clusters=2, random_state=0)

# fit the k-means model to the data
kmeans.fit(X)

# get the cluster assignments for each data point
cluster_labels = kmeans.labels_

# count the number of data points from X[0] to X[2398] that belong to each cluster
c1_count = sum(cluster_labels[:2399] == 0)
c2_count = sum(cluster_labels[:2399] == 1)

print(f"{c1_count} data points belong to c1, and {c2_count} data points belong to c2 out of all positive datapoints")


630 data points belong to c1, and 1769 data points belong to c2 out of all positive datapoints


here we are seeing that out of 2399 datapoints 630 belongs to c1(0) and 1769 belongs to c2(1) cluster. now we can say that 
there are noisy data so clustering will not be very benificial. But on the basis of majority vote we can say that cluster c2 contain positive data points and c1 contains negative datapoints.

In [378]:
cluster_labels

array([1, 0, 0, ..., 1, 1, 1])

In [379]:
#in our dataset we have y =1 or -1 but in clustering c1(0) means negative datapoints so for comparing i changed c1 0 to -1.
cluster_labels[cluster_labels == 0] = -1

In [381]:
cm = confusion_matrix(y, cluster_labels )

# print the confusion matrix
print(cm)

[[  92 4466]
 [ 660 2087]]


In [382]:
import numpy as np
from sklearn.metrics import accuracy_score, f1_score, roc_auc_score, roc_curve



# Calculate accuracy
accuracy = accuracy_score(y, cluster_labels)

# Calculate F1 score
f1 = f1_score(y, cluster_labels)

# Calculate AUC-ROC score
auc_roc = roc_auc_score(y, cluster_labels)

# Get ROC curve values
#fpr, tpr, thresholds = roc_curve(actual_label, f_pred)

print("Accuracy:", accuracy)
print("F1 score:", f1)
print("AUC-ROC score:", auc_roc)

Accuracy: 0.29828884325804245
F1 score: 0.4488172043010753
AUC-ROC score: 0.38996109362113973


Conclusion:
This is my idea to use clustering for this type problem. but i think this approach will work well if dataset is not noisy. in our case there is noisy so this approach is not working good.
Or we can use such clustering that will not be more effected by noisy datapoints like DBSCAN.

In [384]:
from sklearn.cluster import DBSCAN



# initialize DBSCAN with eps and min_samples parameters
dbscan = DBSCAN(eps=0.5, min_samples=)

# fit the DBSCAN model to the data
dbscan.fit(X)

# get the cluster assignments for each data point
cluster_labels1 = dbscan.labels_

# count the number of data points from X[0] to X[2398] that belong to each cluster
c1_count = sum(cluster_labels1[:2399] == 0)
c2_count = sum(cluster_labels1[:2399] == 1)

print(f"{c1_count} data points belong to c1, and {c2_count} data points belong to c2")


2 data points belong to c1, and 1 data points belong to c2
